src/System.Management.Automation/engine/SessionStateContainer.cs

```diff
@@ -3459,86 +3458,56 @@ internal Collection<PSObject> NewItem(string[] paths, string name, string type,
                                 out providerInstance);
                 }
 
+                // Don't support 'New-Item -Type Directory' on the Function provider
+                // if the runspace has ever been in constrained language mode, as the mkdir
+                // function can be abused
+                if (context.ExecutionContext.HasRunspaceEverUsedConstrainedLanguageMode &&
+                    (providerInstance is Microsoft.PowerShell.Commands.FunctionProvider) &&
+                    (string.Equals(type, "Directory", StringComparison.OrdinalIgnoreCase)))
+                {
+                    throw
+                        PSTraceSource.NewNotSupportedException(SessionStateStrings.DriveCmdletProvider_NotSupported);
+                }
+
+                // Symbolic link targets are allowed to not exist on both Windows and Linux
+                bool allowNonexistingPath = false;
+                bool isSymbolicJunctionOrHardLink = false;
+
+                if (type != null)
+                {
+                    WildcardPattern typeEvaluator = WildcardPattern.Get(type + "*", WildcardOptions.IgnoreCase | WildcardOptions.Compiled);
```

Comment:
I don't understand the use of the wildcardpattern here, `type` is a string for filesystem that should be `symboliclink`, `junction`, `hardlink`, etc... right?  why isn't this a switch statement?